File: CodeForces/Python-Forces/C-Closest-to-the-Right.py
###########################################################################################################################
# Given an array of 𝑛 numbers, sorted in non-decreasing order, and 𝑘 queries. 
# For each query, print the minimum index of an array element not less than the given one.
# 
# Input
# The first line of the input contains integers 𝑛 and 𝑘 (0 < 𝑛, 𝑘 ≤ 10^5 ), the length of the a
# rray and the number of queries. The second line contains 𝑛
#  elements of the array, sorted in non-decreasing order. The third line contains 𝑘 
# queries. All array elements and queries are integers, each of which does not exceed 2⋅10^9 in absolute value.
# 
# Output
# For each of the 𝑘 queries, print the minimum index of an array element not less than the given one. 
# If there are none, print 𝑛+1.
# 
# Example input:
# 5 5
# 3 3 5 8 9
# 2 4 8 1 10
# 
# Example output:
# 1
# 3
# 4
# 1
# 6
# 
###########################################################################################################################

from typing import List
import bisect

def read_input() -> (List[int], List[int]):
    n, k = map(int, input().split())
    A = list(map(int, input().split()))
    Q = list(map(int, input().split()))
    return A, Q

def binary_search_bisect(A: List[int], Q: List[int]) -> None:
    min_indices = []
    for q in Q:
        index = bisect.bisect_left(A, q)
        if index < len(A):
            min_indices.append(index + 1)
        else:
            min_indices.append(len(A) + 1)
    
    for num in min_indices:
        print(num)
def binary_search(A: List[int], Q: List[int]) -> None:
    min_indices = []
    for q in Q:
        l, r = 0, len(A) - 1
        result = len(A) + 1
        while l <= r:
            m = l + (r - l) // 2
            if A[m] >= q:
                result = m + 1  # Store the 1-based index
                r = m - 1
            else:
                l = m + 1
        min_indices.append(result)
    
    for num in min_indices:
        print(num)
    
if __name__ == "__main__":
    A, Q = read_input()
    binary_search(A, Q)

File: CodeForces/CPP-Forces/Binary-Search-CodeForces/C-Closest-to-the-Right.cpp
//////////////////////////////////////////////////////////////////////////////////////////////////////////
// Given an array of 𝑛 numbers, sorted in non-decreasing order, and 𝑘 queries. 
// For each query, print the minimum index of an array element not less than the given one.
// 
// Input
// The first line of the input contains integers 𝑛 and 𝑘 (0 < 𝑛, 𝑘 ≤ 10^5 ), the length of the a
// rray and the number of queries. The second line contains 𝑛
//  elements of the array, sorted in non-decreasing order. The third line contains 𝑘 
// queries. All array elements and queries are integers, each of which does not exceed 2⋅10^9 in absolute value.
// 
// Output
// For each of the 𝑘 queries, print the minimum index of an array element not less than the given one. 
// If there are none, print 𝑛+1.
// 
// Example input:
// 5 5
// 3 3 5 8 9
// 2 4 8 1 10
// 
// Example output:
// 1
// 3
// 4
// 1
// 6
// 
//////////////////////////////////////////////////////////////////////////////////////////////////////////

#include <iostream>
#include <vector>

void readInput(int& n, int& k, std::vector<int>& A, std::vector<int>& Q) {
    std::cin >> n >> k;
    A.resize(n);
    Q.resize(k);

    for (int i = 0; i < n; i++) {
        std::cin >> A[i];
    }

    for (int j = 0; j < k; j++) {
        std::cin >> Q[j];
    }
}

void binary_search_bisect(std::vector<int>& A, std::vector<int>& Q) {
    std::vector<int> min_indices;
    for (int i = 0; i < Q.size(); i++) {
        auto index = std::lower_bound(A.begin(), A.end(), Q[i]) - A.begin();
        if (index < A.size()) {
            min_indices.push_back(index + 1);
        } else {
            min_indices.push_back(A.size() + 1);
        }
    }
    for (const int& index : min_indices) {
        std::cout << index << std::endl;
    }
}

int main() {

    int n, k;
    std::vector<int> A, Q;
    readInput(n, k, A, Q);
    std::cout << std::endl;
    binary_search_bisect(A, Q);
    return 0;
}